tests/test_raytracing.py

```diff
@@ -0,0 +1,319 @@
+      positions = (
+          data["0/Position"][:] + np.array([
+              0.5 / self.header["nc"] * self.header["box_size_mpch"],
+              0.5 / self.header["nc"] * self.header["box_size_mpch"], 0
+          ],
+                                           dtype=np.float32)) * self.Mpc_over_h
+    else:
+      positions = data["0/Position"][first:last] * self.Mpc_over_h
+
+    #Enforce periodic boundary conditions
+    for n in (0, 1):
+      positions[:, n][positions[:, n] < 0] += self.header["box_size"]
+      positions[:, n][
+          positions[:, n] > self.header["box_size"]] -= self.header["box_size"]
+
+    #Maybe save
+    if save:
+      self.positions = positions
+
+    #Initialize useless attributes to None
+    self.weights = None
+    self.virial_radius = None
+    self.concentration = None
+
+    #Return
+    return positions
+
+
+def test_density_plane(return_results=False):
+  """ Tests cutting density planes from snapshots against lenstools
+  """
+  klin = np.loadtxt(data_path + '/flowpm/data/Planck15_a1p00.txt').T[0]
+  plin = np.loadtxt(data_path + '/flowpm/data/Planck15_a1p00.txt').T[1]
+  ipklin = iuspline(klin, plin)
+
+  cosmo = flowpm.cosmology.Planck15()
+
+  a0 = 0.9
+  r0 = flowpm.background.rad_comoving_distance(cosmo, a0)
+
+  # Create a state vector
+  initial_conditions = flowpm.linear_field(nc, bs, ipklin, batch_size=2)
+  state = flowpm.lpt_init(cosmo, initial_conditions, a0)
+
+  # Export the snapshot
+  flowpm.io.save_state(cosmo,
+                       state,
+                       a0, [nc, nc, nc], [bs, bs, bs],
+                       'snapshot_density_testing',
+                       attrs={'comoving_distance': r0})
+
+  # Reload the snapshot with lenstools
+  snapshot = FlowPMSnapshot.open('snapshot_density_testing')
+
+  # Cut a lensplane in the middle of the volume
+  lt_plane, resolution, NumPart = snapshot.cutPlaneGaussianGrid(
+      normal=2,
+      plane_resolution=plane_resolution,
+      center=(bs / 2) * snapshot.Mpc_over_h,
+      thickness=(bs / 4) * snapshot.Mpc_over_h,
+      left_corner=np.zeros(3) * snapshot.Mpc_over_h,
+      smooth=None,
+      kind='density')
+
+  # Cut the same lensplane with flowpm
+  fpm_plane = flowpm.raytracing.density_plane(
+      state,
+      nc,
+      center=nc / 2,
+      width=nc / 4,
+      plane_resolution=plane_resolution)
+
+  # Apply additional normalization terms to match lenstools definitions
+  constant_factor = 3 / 2 * cosmo.Omega_m * (constants.H0 / constants.c)**2
+  density_normalization = bs / 4 * r0 / a0
+  fpm_plane = fpm_plane * density_normalization * constant_factor
+
+  # Checking first the mean value, which accounts for any normalization
+  # issues
+  assert_allclose(np.mean(fpm_plane[0]), np.mean(lt_plane), rtol=1e-5)
+
+  # To check pixelwise difference, we need to do some smoothing as lenstools and
+  # flowpm use different painting kernels
+  smooth_lt_plane = np.fft.ifft2(fourier_gaussian(np.fft.fft2(lt_plane),
```

Comment:
Ah yes :-) This is part of the smoothting things we should discuss. Ok, long storry short, I realized that when we want to compute the power spectrum of the smoothed field, if we do the gaussian smoothing in pixel space, the power spectrrum is not very clean (it doesn't exactly go to 0 at high frequencies). But if we do an equivalent smoothing in Fourrier space, the PS looks good, and that's expected because we directly operate in the Fourier domain.

We should talk about all of this more, maybe during a learning-slot ;-)

Reply:
All right,  I'll store my question for the learning-slot! :)